cmd/dockerd/hack/malformed_host_override_test.go

```diff
@@ -0,0 +1,214 @@
+	}
+	expected := []byte("GET /foo\nHost: \r\nConnection: close\r\nUser-Agent: Docker\r\n\r\n")
+	if !bytes.Equal(read[:n], expected) {
+		t.Fatalf("\n%v\n%v\n", read[:n], expected)
+	}
+	l.Close()
+	<-done
+}
+
+func TestHeaderOverrideHackNoOverrideHeaders(t *testing.T) {
+	client, srv := net.Pipe()
+	req := []byte("GET /foo\nHost: /var/run/docker.sock\nUser-Agent: Docker\nFoo: Bar\r\n")
+	read := make([]byte, 4096)
+	done := make(chan struct{})
+
+	l := MalformedHostHeaderOverrideConn{client, true}
+	go func() {
+		for {
+			if _, err := srv.Write(req); err != nil {
+				srv.Close()
+				break
+			}
+		}
+		close(done)
+	}()
+
+	n, err := l.Read(read)
+	if err != nil && err != io.EOF {
+		t.Fatalf("read: %d - %d, err: %v\n%s", n, len(req), err, string(read[:n]))
+	}
+	expected := []byte("GET /foo\nHost: \r\nConnection: close\r\nUser-Agent: Docker\nFoo: Bar\r\n")
+	if !bytes.Equal(read[:n], expected) {
+		t.Fatalf("\n%v\n%v\n", read[:n], expected)
+	}
+	l.Close()
+	<-done
+}
+
+func TestHeaderOverrideHackWithBody(t *testing.T) {
+	client, srv := net.Pipe()
+	req := []byte("GET /foo\nHost: /var/run/docker.sock\nUser-Agent: Docker\r\n\r\ntest something!")
+	read := make([]byte, 4096)
+	done := make(chan struct{})
+
+	l := MalformedHostHeaderOverrideConn{client, true}
+	go func() {
+		for {
+			if _, err := srv.Write(req); err != nil {
+				srv.Close()
+				break
+			}
+		}
+		close(done)
+	}()
+
+	n, err := l.Read(read)
+	if err != nil && err != io.EOF {
+		t.Fatalf("read: %d - %d, err: %v\n%s", n, len(req), err, string(read[:n]))
+	}
+	expected := []byte("GET /foo\nHost: \r\nConnection: close\r\nUser-Agent: Docker\r\n\r\ntest something!")
+	if !bytes.Equal(read[:n], expected[:n]) {
+		t.Fatalf("\n%v\n%v\n", read[:n], expected[:n])
+	}
+	l.Close()
+	<-done
+}
+
+func TestHeaderOverrideHackWithLargeBody(t *testing.T) {
+	client, srv := net.Pipe()
+	req := []byte("GET /foo\nHost: /var/run/docker.sock\nUser-Agent: Docker\r\n\r\ntest something! " + strings.Repeat("test", 15000))
+	read := make([]byte, 4096)
+	done := make(chan struct{})
+
+	l := MalformedHostHeaderOverrideConn{client, true}
+	go func() {
+		for {
+			if _, err := srv.Write(req); err != nil {
+				srv.Close()
+				break
+			}
+		}
+		close(done)
+	}()
+
+	n, err := l.Read(read)
+	if err != nil && err != io.EOF {
+		t.Fatalf("read: %d - %d, err: %v\n%s", n, len(req), err, string(read[:n]))
+	}
+	expected := []byte("GET /foo\nHost: \r\nConnection: close\r\nUser-Agent: Docker\r\n\r\ntest something! " + strings.Repeat("test", 15000))
+	if !bytes.Equal(read[:n], expected[:n]) {
+		t.Fatalf("\n%v\n%v\n", read[:n], expected[:n])
+	}
+	l.Close()
+	<-done
+}
+
+func TestHeaderOverrideHackEnsureDockerClientsOnlyOverride(t *testing.T) {
+	client, srv := net.Pipe()
+	req := []byte("GET /foo\nFoo: Bar\nHost: /var/run/docker.sock\nUser-Agent: Docker\r\n\r\n")
+	read := make([]byte, 4096)
+	done := make(chan struct{})
+
+	l := MalformedHostHeaderOverrideConn{client, true}
+	go func() {
```

Comment:
Yeah, I'm sure we need the goroutine since it's an io.Pipe underneath, but no need for the loop since we are only reading 1 time, and no need to signal done because reads/writes are synchronous